### ISSUE TITLE
matmul kernels for the `i8` case on x86

### ISSUE THREAD

**Issue by bjacob** [openxla/iree#11927]
As of #11849 we have the right tile sizes selected for matmul on x86, so it's down to "bugs" if performance isn't great.

The generated matmul kernels in the `f32` case are actually [good](https://gist.github.com/bjacob/a1e333ec090a824c09799282a5cf84f4).  It's an easy case though. Even AVX-512 offers no more than plain vector mul-adds so the data tiling is set to generate an outer-product kernel, which is the one thing that works well in upstream vector.contract lowerings.

The `i8` case is bad because the instructions that x86 has for this are dot-product instructions (2D dotprods all the way back to MMX (!); 4D dotprods in AVX-512-VNNI) and upstream codegen doesn't know how to use them.

Ideally one would teach that trick to upstream codegen, but as a stepping stone towards that and to get a quicker ROI it makes sense to start in `VectorContractCustomKernels`.

Those custom kernels may be either inline asm or intrinsics. This week I gave a try to the inline asm way and ran into these issues:
1. Spurious register allocations errors, `error: inline assembly requires more registers than available`, bogus as my inline asm statement used only ~22 out of 32 `zmm` registers.
2. matmul kernels need broadcast-multiplication instructions. x86 's multiplication instructions only do broadcasting of memory operands (unlike ARM multiplication instructions which can broadcast a lane of a SIMD register). When I write a lowering of a vector.contract to inline asm, data is already loaded into a SIMD register, it's too late to utilize those instructions with memory operands. The targeting of instructions with memory operands is normally achieved as a peephole optimization in the backend, rewriting (load vector, broadcast one lane onto another vector, multiply with that vector) to (multiply with memory operand). Using inline asm prevents benefiting from this peephole optimization.

TLDR: inline asm doesn't work as well on x86 as it does on ARM.

So the next time that I spend time on this, I will try the intrinsics approach. There is already an example of an intrinsics kernel in VectorContractCustomKernels, but it's targeting the ArmNeon dialect. Here we would have to target the X86Vector dialect. Difficulties (which were why I tried inline asm first before falling back on intrinsics):
1. The X86Dialect does not readily have the instructions that we need, so it would need to be grown as needed. It's upstream, so there will be the usual friction from working across repos etc.  (Just noting this to help estimate the effort).
2. These kernels will be somewhat nontrivial due to inadequacy of x86 instruction set, so there will be plenty of opportunity for the usual "the compiler backend pooped its pants again with intrinsics".
3. The [VNNI instruction](https://www.felixcloutier.com/x86/vpdpbusd) that we would want to use is surprisingly bad, even by x86 standards: it can only multiply this one combination of signednesses: `uint8 * int8`. Of course, what we have is `int8 * int8`. This means that to use this, we would need to xor the top bit (effectively converting `int8` to `uint8` by adding +128) and make up for it in additional arithmetic.  In the past, in packing-matmul libraries such as ruy, this kind of fix-up was implemented for the opposite reason (supporting old `uint8 * uint8` quantized models on newer `int8 * int8` kernels) and that was done at negligible cost thanks to doing the xor-ing in the packing code and treating the fix-up arithmetic as simply an adjustment to the zero-point. But in terms of our MLIR stack here, this would mean handling that all the way back to the original `linalg.quantized_matmul` --- here we are way past that point. We didn't do something wrong here, everyone has standardized on `int8*int8`. Except, I just realized, Intel baked `uint8 * int8` in VNNI and supports no other combination. The only possible rationale that I can think of is someone thought "hey, weights are signed, but activations are surely unsigned because RELU!!!". Anyway, the VNNI kernel is going to be ugly. It may be simpler to use the other VNNI instruction, `VPDPWSSD`, which is `int16 * int16` so only half the throughput, but more guaranteed to be a win over non-VNNI.  Oh and the reason why it'll be a win is that non-VNNI only has non-accumulating dotproducts, so `VPDPWSSD` is better because it's dot-product-and-accumulate, saving the additions.

Another take-away is ukernels will out-perform vector.contract codegen on these use cases because they'll allow us to use (out of line) assembly and because owning the whole loop there will allow us to deal with those fixups more efficiently.

**Comment by allieculp** [comment]
@bjacob Curious if you are working on this, or if it is more a P2/3 item?

**Comment by bjacob** [comment]
x86 ukernels for matmul, including the `i8` case, have landed now, and we are on track to make ukernels available in codegen. This reduces the importance of fixing this in pure codegen, so we can make this P2/P3 now, or even close this for now. Defaulting to P3 because "nice to have" sounds fair.
